src/python/pants/engine/legacy/graph.py

```diff
@@ -677,6 +701,20 @@ def _eager_fileset_with_spec(
   return SourcesSnapshots(snapshots)
 
 
+@rule
+async def sources_snapshots_from_filesystem_specs(
+  filesystem_specs: FilesystemSpecs, glob_match_error_behavior: GlobMatchErrorBehavior,
+) -> SourcesSnapshots:
+  """Resolve the snapshot associated with the provided filesystem specs."""
+  snapshot = await Get[Snapshot](
+    PathGlobs(
+      include=(fs_spec.glob for fs_spec in filesystem_specs),
+      glob_match_error_behavior=glob_match_error_behavior,
+    )
+  )
+  return SourcesSnapshots([SourcesSnapshot(snapshot)])
```

Comment:
This is literally all it took to get `cloc2` and `validate` to work with filesystem specs. V2 is so cool!!!